gio/src/subclass/async_initable.rs

```diff
@@ -0,0 +1,522 @@
+        .expect("Result is not a Rust AsyncInitableResult");
+    if !res.is_valid(initable.as_ref()) {
+        panic!("AsyncInitable is not valid for source object");
+    }
+
+    match res.propagate_error() {
+        None => glib::ffi::GTRUE,
+        Some(e) => {
+            if !error.is_null() {
+                *error = e.into_raw();
+            }
+            glib::ffi::GFALSE
+        }
+    }
+}
+
+// A GAsyncResult that only holds the object and a GError, because GTask is too heavy
+glib::wrapper! {
+    struct AsyncInitableResult(ObjectSubclass<imp::AsyncInitableResult>)
+        @implements AsyncResult;
+}
+
+mod imp {
+    use crate::prelude::*;
+    use crate::subclass::prelude::*;
+    use crate::AsyncInitable;
+
+    impl super::AsyncInitableResult {
+        #[inline]
+        pub(super) fn new(
+            source_object: &AsyncInitable,
+            user_data: glib::Pointer,
+            source_tag: glib::Pointer,
+            error: Option<glib::Error>,
+        ) -> Self {
+            let res = glib::Object::new::<Self>(&[]).unwrap();
+            res.imp()
+                .inner
+                .set(Inner {
+                    source_object: source_object.clone(),
+                    user_data,
+                    source_tag,
+                })
+                .unwrap();
+            res.imp().error.replace(error);
+            res
+        }
+        #[inline]
+        pub(super) fn is_valid(&self, source_object: &AsyncInitable) -> bool {
+            self.imp().inner.get().unwrap().source_object == *source_object
+        }
+        #[inline]
+        pub(super) fn propagate_error(&self) -> Option<glib::Error> {
+            self.imp().error.take()
+        }
+    }
+
+    #[derive(Debug)]
+    struct Inner {
+        source_object: AsyncInitable,
+        user_data: glib::Pointer,
+        source_tag: glib::Pointer,
+    }
+
+    #[derive(Default)]
+    pub(super) struct AsyncInitableResult {
+        inner: once_cell::sync::OnceCell<Inner>,
+        error: std::cell::RefCell<Option<glib::Error>>,
+    }
+
+    #[glib::object_subclass]
+    impl ObjectSubclass for AsyncInitableResult {
+        const NAME: &'static str = "GRustAsyncInitableResult";
+        type Type = super::AsyncInitableResult;
+        type Interfaces = (crate::AsyncResult,);
+    }
+
+    impl ObjectImpl for AsyncInitableResult {}
+
+    impl AsyncResultImpl for AsyncInitableResult {
+        #[inline]
+        fn user_data(&self, _res: &Self::Type) -> glib::Pointer {
+            self.inner.get().unwrap().user_data
+        }
+        #[inline]
+        fn source_object(&self, _res: &Self::Type) -> Option<glib::Object> {
+            Some(self.inner.get().unwrap().source_object.clone().upcast())
+        }
+        #[inline]
+        fn is_tagged(&self, _res: &Self::Type, source_tag: glib::Pointer) -> bool {
+            self.inner.get().unwrap().source_tag == source_tag
+        }
+    }
+}
+
+#[cfg(test)]
```

Comment:
Did you check if the tests are valgrind clean?

Reply:
Yes, I only get these leaks about type data, which I was also getting before with other types:

```
==2685840==
==2685840== HEAP SUMMARY:
==2685840==     in use at exit: 48,608 bytes in 345 blocks
==2685840==   total heap usage: 1,411 allocs, 1,066 frees, 159,656 bytes allocated
==2685840==
==2685840== 32 bytes in 1 blocks are possibly lost in loss record 145 of 317
==2685840==    at 0x483F7B5: malloc (vg_replace_malloc.c:381)
==2685840==    by 0x17B0D4: glib::subclass::types::interface_init (in /gtk-rs-core/target/release/deps/gio-95981b62c749adcd)
==2685840==    by 0x48BCFF8: type_iface_vtable_iface_init_Wm (gtype.c:2125)
==2685840==    by 0x48BF7EA: type_class_init_Wm (gtype.c:2323)
==2685840==    by 0x48BF7EA: g_type_class_ref (gtype.c:3012)
==2685840==    by 0x1C8BF9: glib::object::Object::with_type (in /gtk-rs-core/target/release/deps/gio-95981b62c749adcd)
==2685840==    by 0x14E3DC: glib::object::Object::new (in /gtk-rs-core/target/release/deps/gio-95981b62c749adcd)
==2685840==    by 0x1645FA: gio::async_initable::<impl gio::auto::async_initable::AsyncInitable>::new_future (in /gtk-rs-core/target/release/deps/gio-95981b62c749adcd)
==2685840==    by 0x145F74: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll (in /gtk-rs-core/target/release/deps/gio-95981b62c749adcd)
==2685840==    by 0x1CB57E: glib::main_context::<impl glib::auto::main_context::MainContext>::with_thread_default (in /gtk-rs-core/target/release/deps/gio-95981b62c749adcd)
==2685840==    by 0x1CC299: _ZN4glib20main_context_futures10TaskSource8dispatch17hfc5962ecab52f8d6E.llvm.10211790655316037810 (in /gtk-rs-core/target/release/deps/gio-95981b62c749adcd)
==2685840==    by 0x4B28F8A: g_main_dispatch (gmain.c:3417)
==2685840==    by 0x4B28F8A: g_main_context_dispatch (gmain.c:4135)
==2685840==    by 0x4B29237: g_main_context_iterate.constprop.0 (gmain.c:4211)
==2685840==
==2685840== 40 bytes in 1 blocks are possibly lost in loss record 175 of 317
==2685840==    at 0x483F7B5: malloc (vg_replace_malloc.c:381)
==2685840==    by 0x17A984: glib::subclass::types::interface_init (in /gtk-rs-core/target/release/deps/gio-95981b62c749adcd)
==2685840==    by 0x48BCFF8: type_iface_vtable_iface_init_Wm (gtype.c:2125)
==2685840==    by 0x48BF7EA: type_class_init_Wm (gtype.c:2323)
==2685840==    by 0x48BF7EA: g_type_class_ref (gtype.c:3012)
==2685840==    by 0x48A765F: g_object_newv (gobject.c:2224)
==2685840==    by 0x1C988F: glib::object::Object::new_internal (in /gtk-rs-core/target/release/deps/gio-95981b62c749adcd)
==2685840==    by 0x1C8D51: glib::object::Object::with_type (in /gtk-rs-core/target/release/deps/gio-95981b62c749adcd)
==2685840==    by 0x14F35C: glib::object::Object::new (in /gtk-rs-core/target/release/deps/gio-95981b62c749adcd)
==2685840==    by 0x143C6D: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll (in /gtk-rs-core/target/release/deps/gio-95981b62c749adcd)
==2685840==    by 0x1CB57E: glib::main_context::<impl glib::auto::main_context::MainContext>::with_thread_default (in /gtk-rs-core/target/release/deps/gio-95981b62c749adcd)
==2685840==    by 0x1CC299: _ZN4glib20main_context_futures10TaskSource8dispatch17hfc5962ecab52f8d6E.llvm.10211790655316037810 (in /gtk-rs-core/target/release/deps/gio-95981b62c749adcd)
==2685840==    by 0x4B28F8A: g_main_dispatch (gmain.c:3417)
==2685840==    by 0x4B28F8A: g_main_context_dispatch (gmain.c:4135)
==2685840==
==2685840== 84 bytes in 1 blocks are possibly lost in loss record 229 of 317
==2685840==    at 0x483F7B5: malloc (vg_replace_malloc.c:381)
==2685840==    by 0x1393EC: hashbrown::raw::RawTable<T,A>::reserve_rehash (in /gtk-rs-core/target/release/deps/gio-95981b62c749adcd)
==2685840==    by 0x1398BA: hashbrown::raw::RawTable<T,A>::insert (in /gtk-rs-core/target/release/deps/gio-95981b62c749adcd)
==2685840==    by 0x17B1F8: glib::subclass::types::interface_init (in /gtk-rs-core/target/release/deps/gio-95981b62c749adcd)
==2685840==    by 0x48BCFF8: type_iface_vtable_iface_init_Wm (gtype.c:2125)
==2685840==    by 0x48BF7EA: type_class_init_Wm (gtype.c:2323)
==2685840==    by 0x48BF7EA: g_type_class_ref (gtype.c:3012)
==2685840==    by 0x1C8BF9: glib::object::Object::with_type (in /gtk-rs-core/target/release/deps/gio-95981b62c749adcd)
==2685840==    by 0x14E3DC: glib::object::Object::new (in /gtk-rs-core/target/release/deps/gio-95981b62c749adcd)
==2685840==    by 0x1645FA: gio::async_initable::<impl gio::auto::async_initable::AsyncInitable>::new_future (in /gtk-rs-core/target/release/deps/gio-95981b62c749adcd)
==2685840==    by 0x145F74: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll (in /gtk-rs-core/target/release/deps/gio-95981b62c749adcd)
==2685840==    by 0x1CB57E: glib::main_context::<impl glib::auto::main_context::MainContext>::with_thread_default (in /gtk-rs-core/target/release/deps/gio-95981b62c749adcd)
==2685840==    by 0x1CC299: _ZN4glib20main_context_futures10TaskSource8dispatch17hfc5962ecab52f8d6E.llvm.10211790655316037810 (in /gtk-rs-core/target/release/deps/gio-95981b62c749adcd)
==2685840==
==2685840== 84 bytes in 1 blocks are possibly lost in loss record 230 of 317
==2685840==    at 0x483F7B5: malloc (vg_replace_malloc.c:381)
==2685840==    by 0x1393EC: hashbrown::raw::RawTable<T,A>::reserve_rehash (in /gtk-rs-core/target/release/deps/gio-95981b62c749adcd)
==2685840==    by 0x1398BA: hashbrown::raw::RawTable<T,A>::insert (in /gtk-rs-core/target/release/deps/gio-95981b62c749adcd)
==2685840==    by 0x17AAB3: glib::subclass::types::interface_init (in /gtk-rs-core/target/release/deps/gio-95981b62c749adcd)
==2685840==    by 0x48BCFF8: type_iface_vtable_iface_init_Wm (gtype.c:2125)
==2685840==    by 0x48BF7EA: type_class_init_Wm (gtype.c:2323)
==2685840==    by 0x48BF7EA: g_type_class_ref (gtype.c:3012)
==2685840==    by 0x48A765F: g_object_newv (gobject.c:2224)
==2685840==    by 0x1C988F: glib::object::Object::new_internal (in /gtk-rs-core/target/release/deps/gio-95981b62c749adcd)
==2685840==    by 0x1C8D51: glib::object::Object::with_type (in /gtk-rs-core/target/release/deps/gio-95981b62c749adcd)
==2685840==    by 0x14F35C: glib::object::Object::new (in /gtk-rs-core/target/release/deps/gio-95981b62c749adcd)
==2685840==    by 0x143C6D: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll (in /gtk-rs-core/target/release/deps/gio-95981b62c749adcd)
==2685840==    by 0x1CB57E: glib::main_context::<impl glib::auto::main_context::MainContext>::with_thread_default (in /gtk-rs-core/target/release/deps/gio-95981b62c749adcd)
```

Reply:
Yeah those are expected one-time leaks